### ISSUE TITLE
integrate with pacs server

### ISSUE THREAD

**Issue by veezo2007pk** [knopkem/dicomweb-pacs#61]
i need to connect with pacs server any option in this project?i saw thats its studying data from  data folder which is in local, but i have pacs server which storing data in different way

**Comment by knopkem** [comment]
Hi, yes you can push from other PACs to this one to import images. In fact it's the only way. The data folder is holding the images that got imported. Have a look at the config and add any peers than use c-store from that peer to push images

**Comment by knopkem** [comment]
It cannot pull images though, if you need this I recommend the dicomweb-proxy project

**Comment by veezo2007pk** [comment]
please guide me how i can import as i am using leadtools pacs server?
i am able to get patient data from pacs server but when i click on any image it gives error as its taking from data and my data folder is empty

**Comment by knopkem** [comment]
Can you describe what you want to achieve. From what you wrote it looks
like you want to visualize (with OHIF viewer) images that are on your lead
tool PACS. If this is the case, dicomweb-pacs would not be the right tool,
dicomweb-proxy would be recommended.

veezo2007pk ***@***.***> schrieb am Mi., 15. März 2023, 09:34:

**Comment by veezo2007pk** [comment]
i have pacs server which is leadtools. I need a web based viewer that can integrate with pacs server. 
const config = {
    
  source: {
    aet: "KIH_PACS_SS",
    ip: "172.16.12.50",
    port: "105"
  },
    peers: [
      {
        aet: "DICOMWEB_PACS",
        ip: "172.16.4.99",
        port: "5001"
      }],
      here is my code source is my pacs server, and peer is my local. Now issue i am facing i am able to get patient data but when i click on patient and want to see any study its giving me error, because its taking data from data folder

**Comment by knopkem** [comment]
Ok, I see. First of all, you got it inverted:
source defines the naming and port of the dicomweb-pacs (This usually needs to be added to the list of known PACS in the other tools (PACS, Workstations etc.) that want to communicate.
Peers is a list of communication partners dicomweb accepts and knows, so this would be your lead tool server.
However the viewer would show nothing as the dicomweb-pacs project does not query those peers. You would need to explicitly push images to the dicomweb-pacs in order to view them, which is probably not what you want.
 
What you want is a proxy, a tool that just converts your dicomweb requests (coming from the viewer) to "PACS speak" and converts data on the fly. This is what https://github.com/knopkem/dicomweb-proxy was made for. So try this one. You might get the same issue though, that you can find patiens in the study browser but once you click on them the viewer shows nothing. In this case, check your lead tool pacs to know and accept the proxy, so it forwards the images correctly.

You should also double check if you have access to this option: https://www.leadtools.com/help/sdk/v21/dh/to/leadtools-dicomweb-services.html
If yes, than I highly recommend to directly use the OHIF viewer with leadtool as there would be no need to proxy the requests (leadtool "understands" the requests from the viewer natively). This would be way faster. For how to do this please consult the OHIF project: https://github.com/OHIF/Viewers

Good luck

**Comment by veezo2007pk** [comment]
this is my pacs server setting
 "aet": "KIH_PACS_SS",
    "ip": "172.16.12.50",
    "port": 501
    
    and this is setting of my pc from where i am running dicomweb-proxy
    172.16.5.91
    5000
    i dont know aet of my laptop?
    
    i am still confused how to do

**Comment by knopkem** [comment]
You should read about how communication with DIMSE works in the medical world.

This is how your config (part) in dicomweb-proxy should look like:

  "source": {
    "aet": "DICOMWEB_PROXY",
    "ip": "172.16.5.91",
    "port": 8888
  },

  "peers": [
    {
      "aet": "KIH_PACS_SS",
      "ip": "172.16.12.50",
      "port": 501
    }
  ],

For the source IP you can also use 0.0.0.0 (means listening on all interfaces), however in your LeadTool config you might need to register the DICOMWEB-PROXY as a known SCU. It might also work already because the default way to fetch data is C-GET, but depends on the PACS implementation. If you do register it (always recommended) than don't use 0.0.0.0 but the real IP (172.16.5.91).
The AET of you DICOMWEB-PROXY on your laptop is an arbitrary name, you could also choose another name (but there are limits of which chars are allowed and also the length is limited, so best stick with the default). The name is only important for LEADTOOLS to identify the proxy and allow it to communicate. The same goes the other way, the peers are lists (in your case 1) communication partners, with Name (AET), IP and Port to know where to route the requests to.

**Comment by veezo2007pk** [comment]
i have tested DICOMWEB-PROXY yesterday it worked with my pacs server and i see xray study as well, then suddenly my pacs server stopped dont know why? i need to know one thing how DICOMWEB-PROXY is working? it download study data in data folder 1st and then display? can you explain me in detail about this?

**Comment by knopkem** [comment]
Hi, the workflow is as you wrote: the proxy needs to download the images first in order to parse and repack them. The task of the proxy is to translate between two different protocols: dicomweb <-> dimse. The goal is to be able to connect older PACS servers that don't implement dicomweb protocol with modern DICOM viewers (webbased).
This is why I said that using the native dicomweb protocoll of your leadtool server would be preferred as this is the only way that there is no extra download step needed.
Anyway, back to the proxy: the default download granularity is: SERIES level. This means the viewer requests the first image from a series, than the proxy will first fetch the whole series before sending images back to the viewer. This can also be changed to IMAGE level (each image will be requested and forwarded) but this is  usually much slower as there is a lot of communication overhead now added, but you could try it, by setting in default,js:
{
"useFetchLevel": "IMAGE",
"fullMeta": false, 
}
Hower, this also disabled the option in the viewer to go into 3D mode (MPR) as this requires more data to be parsed.
So my recommendation: try to find out if you can connect the OHIF viewer directly to your pacs. You would just need to configure the app-config.js file (from public folder) to point to the right WADO urls and host the static webapp files using any http-server (apache, ngnix etc. or http-server (npm) etc )

**Comment by veezo2007pk** [comment]
where i can find OHIF viewer open source? i have tried once but i am not able to run it as its giving lot of errors. If i use proxy with fetchlevel :image, any solution for  3D mode (MPR)?

**Comment by knopkem** [comment]
The project is hosted here: https://github.com/OHIF/Viewers, however no need to compile it yourself as you can use the version (should be latest) from this repo (in public folder), just adapt the app-config.js file.
When using the proxy with imediately showing the images, there is unfortunately no way currently for the proxy that the viewer enables 2D. The reason is that the OHIF viewer requires a detailed metadata response (from the proxy) about the full SERIES. Since the proxy only has parts of the series at that time it cannot answer this call.
Personally I would use SERIES level though and see if the speed is still ok for you. If the connection between the proxy and the pacs is good than downloading big series before viewing should not be a problem. Also the second time you open the series it uses the cache from the proxy.

**Comment by knopkem** [comment]
No need to use the OHIF repo, copy the public folder from this repo (https://github.com/knopkem/dicomweb-proxy/tree/master/public) and then configure the file: https://github.com/knopkem/dicomweb-proxy/tree/master/public/app-config.js. This part is important:

```
 servers: {
    dicomWeb: [
      {
        name: 'DICOMWEB_PACS',
        wadoUriRoot: 'wadouri',
        qidoRoot: 'rs',
        wadoRoot: 'rs',
        qidoSupportsIncludeField: true,
        imageRendering: 'wadouri',
        thumbnailRendering: 'wadouri',
        enableStudyLazyLoad: true,
        supportsFuzzyMatching: true,
      },
    ],
  },
```

You can see the routes of your server here: https://www.leadtools.com/help/sdk/v20/dh/to/leadtools-dicomweb-wado-rs.html

So that would be:
```
 servers: {
    dicomWeb: [
      {
        name: 'LEADTOOLS',
        wadoUriRoot: '172.16.12.50/api/uri',
        qidoRoot: '172.16.12.50/api/wado-rs',
        wadoRoot: '172.16.12.50/api/qido-rs',
        qidoSupportsIncludeField: true,
        imageRendering: 'wadouri',
        thumbnailRendering: 'wadouri',
        enableStudyLazyLoad: true,
        supportsFuzzyMatching: true,
      },
    ],
  },
```

Then start a webserver and host the files, e.g. within your public folder
`npx http-server .`
Open a browser on 'localhost:8080' (or whatever port it shows in the console)

**Comment by veezo2007pk** [comment]
This is my server config
IP:172.16.12.50
Port:105
Are title:KIH_pacs_ss
How its gonna work with the config you provided above?

**Comment by knopkem** [comment]
Port 105 is only used for dimes networking. Dicomweb uses rest API calls, so the viewer is connecting to port 80 of your lead tool server. Read the docs about it

**Comment by veezo2007pk** [comment]
i am facing one issue when i connect with leadtool pacs server its working fine and i am able to see study as well, but there is one issue that cpu usage of leadtool server get too high as seen in image i have attached
![WhatsApp Image 2023-03-16 at 15 31 38](https://user-images.githubusercontent.com/36745222/225590281-e8b71fae-54a3-422b-ae28-c4c00067b44f.jpg)
cpu usage gets to 50 60 %

**Comment by knopkem** [comment]
That is not coming from the proxy, also this could be normal, when you open
a study it starts fetching all series in parallel. If compression is used
than this is totally normal. Check with leadtools support.

veezo2007pk ***@***.***> schrieb am Do., 16. März 2023, 11:32:

**Comment by veezo2007pk** [comment]
i am trying to run echo report with dicomeweb proxy, but its giving error not able to run it

**Comment by veezo2007pk** [comment]
can you check this report ultrasound its showing like this
![Untitled](https://user-images.githubusercontent.com/36745222/225844315-0089e662-3fb2-4b6e-968c-044b707a971a.png)
what can be issue?

**Comment by knopkem** [comment]
please open a new issue with this (including an anonymized sample file to reproduce)
